src/pykeen/triples/triples_factory.py

```diff
@@ -163,6 +169,291 @@ def _map_triples_elements_to_ids(
+    labeled_column = np.asanyarray(labeled_column)
+    mapped_column = mapper(labeled_column, (unknown_id,))
+    return torch.as_tensor(data=mapped_column, dtype=torch.long)
+
+
+def _label_column(
+    mapped_column: Union[np.ndarray, Sequence[int], torch.LongTensor, int],
+    vectorized_labeler: Callable[[np.ndarray, Tuple[str]], np.ndarray],
+    unknown_label: str,
+) -> np.ndarray:
+    """Apply vectorized labeler to an array of IDs.
+
+    :param mapped_column:
+        The IDs.
+    :param vectorized_labeler:
+        The vectorized labeler.
+    :param unknown_label:
+        A label to use for unknown IDs.
+
+    :return:
+        An array of labels, same shape as mapped_column.
+    """
+    # Input normalization
+    if isinstance(mapped_column, int):
+        mapped_column = [mapped_column]
+    if torch.is_tensor(mapped_column):
+        mapped_column = mapped_column.cpu().numpy()
+    mapped_column = np.asanyarray(mapped_column)
+
+    # Actual labeling
+    return vectorized_labeler(mapped_column, (unknown_label,))
+
+
+@dataclass
+class LabelMapping:
+    """A mapping from labels to IDs."""
+
+    #: The mapping for entities
+    entity_label_to_id: EntityMapping
+
+    #: The mapping for relations
+    relation_label_to_id: RelationMapping
+
+    #: A dictionary mapping each relation to its inverse, if inverse triples were created
+    relation_to_inverse: Optional[Mapping[str, str]] = None
+
+    @staticmethod
+    def from_labeled_triples(triples: LabeledTriples) -> 'LabelMapping':
+        """Create a mapping from labeled triples."""
+        return LabelMapping(
+            entity_label_to_id=create_entity_mapping(triples=triples),
+            relation_label_to_id=create_relation_mapping(relations=triples[:, 1])
+        )
+
+    @property
+    def entity_id_to_label(self) -> Mapping[int, str]:
+        """The mapping from entity IDs to labels."""
+        return invert_mapping(mapping=self.entity_label_to_id)
+
+    @property
+    def relation_id_to_label(self) -> Mapping[int, str]:
+        """The mapping from relation IDs to labels."""
+        return invert_mapping(mapping=self.relation_label_to_id)
+
+    @property
+    def max_entity_id(self) -> int:
+        """Return the maximum entity ID plus 1."""
+        return max(self.entity_label_to_id.values()) + 1
+
+    @property
+    def max_relation_id(self) -> int:
+        """Return the maximum relation ID plus 1."""
+        return max(self.relation_label_to_id.values()) + 1
+
+    @property
+    def is_compact(self) -> bool:
+        """Whether the mapping is compact, i.e. the IDs are consecutive from 0 to num_choices - 1."""
+        return all(
+            set(mapping.values()) == set(range(len(mapping)))
+            for mapping in (self.entity_label_to_id, self.relation_label_to_id)
+        )
+
+    @property
+    def _vectorized_entity_mapper(self) -> Callable[[np.ndarray, Tuple[int]], np.ndarray]:
+        return np.vectorize(self.entity_label_to_id.get)
```

Comment:
this causes the mapping to get generated every time you access the attribute `get` so I guess it makes sense we cache it explicitly instead of using a property

Reply:
or is this set up because it might change?

Reply:
No, it should not change. As before, I am unsure whether it is used often enough and costs enough computation time to justify the memory overhead of storing it additionally.

Reply:
cf. dac6e5c